src/nf_heuristic.py

```diff
@@ -0,0 +1,570 @@
+    option_nf.sort(key = lambda l: l[0])
+    option_set = list(option_nf)
+    
+#    print(option_nf)
+    return module_list
+
+def run_p4_compiler(p4_filename):
+    status = stage_feasible(p4_filename)
+    return status
+
+def generate_code(conf_parser, module_list, p4_filename, p4_version):
+    output_fp = open(p4_filename, 'w')
+    output_fp.write("\n")
+    p4_node_lists = conf_parser.conf_parser_get_global_p4_nodes(copy.deepcopy(module_list))
+#    log_module(module_list)
+    default_nsh_node = nf_node()
+    default_nsh_node.setup_node_from_argument('sys_default', 'SYS', 0, 0)
+    p4_node_lists.insert(0, [default_nsh_node])
+    all_p4_nodes = []
+    for node_list in p4_node_lists:
+        # for each service path, we do topo sorting
+        node_list.sort(cmp=lambda x,y:cmp(x.finish_time, y.finish_time), reverse=True)
+        all_p4_nodes += node_list
+    p4_list = copy.deepcopy(all_p4_nodes)
+
+    # Use 'nfcp_library_parser' to parse each P4 library
+    # Method: libParser.nfcp_lib_parser(lib_repo, nf_node)
+    print("NFCP Lib Parser is running...")
+
+    for p4_node in p4_list:
+        lib_filename = p4_14_module_list[p4_node.nf_class]
+        if p4_version == 'p414':
+            lib_dir = os.path.join(P414_LIB_DIR, lib_filename)
+        elif p4_version == 'p416':
+            lib_dir = os.path.join(P416_LIB_DIR, lib_filename)
+        lib_parser = libParser.nfcp_lib_parser(lib_dir, p4_node)
+
+    p4_generator = codeGenerator.nfcp_code_generator(conf_parser.scanner, p4_list, p4_version)
+    print("NFCP P4 Library Combiner is running...")
+    p4_generator.lib_combine_main()
+
+    print("NFCP P4 Code Generator is running...")
+    output_fp.write(p4_generator.code_generator_main())
+
+    output_fp.close()
+    print("NFCP Compiler Ends!")
+
+def bess_code(conf_parser, all_nodes, final_bess_filename):
+    bess_fp = open(final_bess_filename, 'w')
+    BG.generate_bess(conf_parser, all_nodes)
+
+def fitp4(conf_parser, module_list, p4_filename, p4_version):
+    generate_code(conf_parser, module_list, p4_filename, p4_version)
+    success_flag = run_p4_compiler(p4_filename)
+    return success_flag
+
+def next_placement_to_offload_p4(module_list):
+    global last_index
+    if len(option_nf) != 0:
+        if(last_index!= -1):
+            module_list[last_index].offload = False
+            module_list[last_index].nf_type = 0
+        pair = option_nf.pop(0)
+        last_index = pair[1]
+        module_list[last_index].offload = True
+        module_list[last_index].nf_type = 1
+    elif len(option_set)>0:
+        fix_module = option_set.pop(0)
+        fix_index = fix_module[1]
+        module_list[fix_index].offload = True
+        module_list[fix_index].nf_type = 1
+        option_nf = list(option_set)
+        module_list = next_placement_to_offload_p4(module_list)
+    else:
+        print("NO AVAILABLE PLACEMENT")
+
+    return module_list
+
+def highest_core_allocation_derived_from_LP(module_list):
```

Comment:
Sorry, this function name is a little bit confused to me.
Do you prefer core_alloc_highest_tput_derived_from_LP for a better readability? Or others if you prefer.